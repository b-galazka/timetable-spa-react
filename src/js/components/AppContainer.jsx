import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { withRouter } from 'react-router-dom';
import propTypes from 'prop-types';

import { getInitialState } from '../actions/initialState';
import { getTimetableObject } from '../actions/timetableObject';

import Content from './Content';
import Loader from './Loader';

import urlTranslations from '../../json/urlTranslations';

function mapStateToProps(state) {

    const { fetched } = state.initialState;

    return {
        initialStateFetched: fetched
    };
}

function mapDispachToProps(dispatch) {

    return {
        getTimetableObject: bindActionCreators(getTimetableObject, dispatch),
        getInitialState: bindActionCreators(getInitialState, dispatch)
    };
}

class AppContainer extends Component {

    render() {

        const { initialStateFetched } = this.props;

        if (initialStateFetched) {
            
            return <Content />;
        }

        return <Loader />;
    }

    componentDidMount() {

        this.fetchInitialState();
    }

    componentWillReceiveProps(nextProps) {

        const { props } = this;
        const { slug, type } = props.match.params;
        const { slug: nextSlug, type: nextType } = nextProps.match.params;

        if (slug !== nextSlug || type !== nextType) {

            props.getTimetableObject({
                objectType: urlTranslations[nextType],
                slug: nextSlug
            });
        }
    }

    fetchInitialState() {

        const { props } = this;
        const { slug, type } = props.match.params;

        props.getInitialState({
            slug,
            objectType: urlTranslations[type]
        });
    }
}

AppContainer.propTypes = {

    // redux
    initialStateFetched: propTypes.bool.isRequired,
    getInitialState: propTypes.func.isRequired,
    getTimetableObject: propTypes.func.isRequired,

    // router
    match: propTypes.shape({
        params: propTypes.shape({
            slug: propTypes.string.isRequired,
            type: propTypes.string.isRequired
        }).isRequired
    }).isRequired
};

export default withRouter(connect(mapStateToProps, mapDispachToProps)(AppContainer));