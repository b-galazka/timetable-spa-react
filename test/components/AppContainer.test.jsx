import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-15';
import renderer from 'react-test-renderer';

import AppContainer from '../../src/js/components/AppContainer';

Enzyme.configure({ adapter: new Adapter() });

jest.mock('../../src/js/components/Content', () => 'Content');
jest.mock('../../src/js/components/Loader', () => 'Loader');

jest.mock(
    '../../src/json/urlTranslations',
    () => ({ ABC: 'translated ABC', BCD: 'translated BCD' })
);

describe('AppContainer component', () => {

    let appContainer;

    beforeEach(() => {

        appContainer = shallow(
            <AppContainer.WrappedComponent
                initialStateFetched
                getInitialState={jest.fn()}
                getTimetableObject={jest.fn()}

                match={{
                    params: { slug: '', type: '' }
                }}
            />
        ).instance();

        appContainer.props = { ...appContainer.props };
    });

    it('should render loader if initial data has not been fetched yet', () => {

        const tree = renderer.create(
            <AppContainer.WrappedComponent
                initialStateFetched={false}
                getInitialState={jest.fn()}
                getTimetableObject={jest.fn()}

                match={{
                    params: { slug: '', type: '' }
                }}
            />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should render content if initial data has been already fetched', () => {

        const tree = renderer.create(
            <AppContainer.WrappedComponent
                initialStateFetched
                getInitialState={jest.fn()}
                getTimetableObject={jest.fn()}

                match={{
                    params: { slug: '', type: '' }
                }}
            />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    describe('AppContainer.prototype.componentDidMount', () => {

        it('should call props.getInitialState', () => {

            const getInitialStateMock = jest.fn();

            appContainer.props.match.params = { slug: 'XYZ', type: 'ABC' };
            appContainer.props.getInitialState = getInitialStateMock;

            appContainer.componentDidMount();

            expect(getInitialStateMock).toHaveBeenCalledTimes(1);

            expect(getInitialStateMock).toHaveBeenCalledWith({
                slug: 'XYZ',
                objectType: 'translated ABC'
            });
        });
    });

    describe('AppContainer.prototype.componentWillReceiveProps', () => {

        it('should call props.getTimetableObject ' +
            'if `type` url param has been changed', () => {

            const nextProps = {
                match: {
                    params: { type: 'BCD', slug: 'XYZ' }
                }
            };

            const getTimetableObjectMock = jest.fn();

            appContainer.props.match.params = { slug: 'XYZ', type: 'ABC' };
            appContainer.props.getTimetableObject = getTimetableObjectMock;

            appContainer.componentWillReceiveProps(nextProps);

            expect(getTimetableObjectMock).toHaveBeenCalledTimes(1);

            expect(getTimetableObjectMock).toHaveBeenCalledWith({
                slug: 'XYZ',
                objectType: 'translated BCD'
            });
        });

        it('should call props.getTimetableObject ' +
            'if `slug` url param has been changed', () => {

            const nextProps = {
                match: {
                    params: { type: 'ABC', slug: 'ZYX' }
                }
            };

            const getTimetableObjectMock = jest.fn();

            appContainer.props.match.params = { slug: 'XYZ', type: 'ABC' };
            appContainer.props.getTimetableObject = getTimetableObjectMock;

            appContainer.componentWillReceiveProps(nextProps);

            expect(getTimetableObjectMock).toHaveBeenCalledTimes(1);

            expect(getTimetableObjectMock).toHaveBeenCalledWith({
                slug: 'ZYX',
                objectType: 'translated ABC'
            });
        });

        it('should not call props.getTimetableObject ' +
            'if url has not been updated', () => {

            const nextProps = {
                match: {
                    params: { slug: 'XYZ', type: 'ABC' }
                }
            };

            const getTimetableObjectMock = jest.fn();

            appContainer.props.match.params = { slug: 'XYZ', type: 'ABC' };
            appContainer.props.getTimetableObject = getTimetableObjectMock;

            appContainer.componentWillReceiveProps(nextProps);

            expect(getTimetableObjectMock).not.toHaveBeenCalled();
        });
    });
});