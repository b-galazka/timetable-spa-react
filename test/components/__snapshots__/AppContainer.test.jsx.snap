// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppContainer component should render content if initial data has been already fetched 1`] = `<Content />`;

exports[`AppContainer component should render loader if initial data has not been fetched yet 1`] = `<Loader />`;
